File: desktop/core/protocol.py
"""
Protokol komunikasi TCP antara Android (client) dan Desktop (server).

Frame format:
    [1 byte msg_type][4 byte length (big-endian uint32)][payload]

Message types:
    0x01 FRAME      -> payload = JPEG bytes (Android -> Desktop)
    0x02 TELEMETRY  -> payload = JSON utf-8 (Android -> Desktop)
    0x03 COMMAND    -> payload = JSON utf-8 (Desktop -> Android)
    0x04 PING       -> payload = token string (Desktop -> Android)
    0x05 PONG       -> payload = token string (Android -> Desktop)
    0x06 AUTH       -> payload = JSON {"camera_id":.., "password":..} (Android -> Desktop)
    0x07 AUTH_RESULT-> payload = JSON {"ok": bool} (Desktop -> Android)
"""
import struct

MSG_FRAME = 0x01
MSG_TELEMETRY = 0x02
MSG_COMMAND = 0x03
MSG_PING = 0x04
MSG_PONG = 0x05
MSG_AUTH = 0x06
MSG_AUTH_RESULT = 0x07

HEADER_SIZE = 5  # 1 + 4 bytes
DEFAULT_PORT = 8765


def pack_message(msg_type: int, payload: bytes) -> bytes:
    return struct.pack(">BI", msg_type, len(payload)) + payload


def unpack_header(header: bytes):
    msg_type, length = struct.unpack(">BI", header)
    return msg_type, length

File: android/app/src/main/java/com/kittenguardian/cctv/Protocol.kt
package com.kittenguardian.cctv

import java.io.ByteArrayOutputStream
import java.io.DataOutputStream

/**
 * Format frame: [1 byte msg_type][4 byte length big-endian][payload]
 * Harus sama persis dengan desktop/core/protocol.py
 */
object Protocol {
    const val MSG_FRAME: Int = 0x01
    const val MSG_TELEMETRY: Int = 0x02
    const val MSG_COMMAND: Int = 0x03
    const val MSG_PING: Int = 0x04
    const val MSG_PONG: Int = 0x05
    const val MSG_AUTH: Int = 0x06
    const val MSG_AUTH_RESULT: Int = 0x07

    const val HEADER_SIZE = 5
    const val DEFAULT_PORT = 8765

    fun pack(msgType: Int, payload: ByteArray): ByteArray {
        val out = ByteArrayOutputStream(Protocol.HEADER_SIZE + payload.size)
        val dos = DataOutputStream(out)
        dos.writeByte(msgType)
        dos.writeInt(payload.size) // DataOutputStream writeInt = big-endian
        dos.write(payload)
        return out.toByteArray()
    }
}
